### ISSUE TITLE
source/destination not properly updated when source contents are changed 

### ISSUE THREAD

**Issue by biskit1** [czeckd/angular-dual-listbox#121]
Hi @czeckd,  
First of all, this is an awesome little Module- thank you for making it!  
I had a question about some customization, hoping you can provide some insight. Not sure if the way I'm going about it does not work because of some bug in my code, or because of the behavior of the component, but if you recognize it as being a component thing and can give me some pointers that'd be much appreciated. So here goes...  (this is with using strings)
  
I have a structure like so:  
`parent` has an array `allSourceOptions` which gets passed into dual list component to be used as its `source`. the dual lists `destination` is linked from another array in the parent, lets call it `selectedSourceOptions`  
   
Then I have a child component of `parent`, lets call it `child`.  `child` gets the list of `selectedSourceOptions` from `parent`, which sets `source` in the dual list component (which is inside of the `child`). The `destination` is set from another input that is sent in to `child` from `parent`.   
  
The problem is:  
I choose some options in `parent` which becomes the `source` inside of `child`s dual list.   
In `child`, I choose some of those options to send to destination.  
The problem occurs when I then try to remove some of the `destination` items in the `parent` back into `allSourceOptions`.  
If `child` contains some of those removed options in its `source`, they are removed as expected. But if any of the removed items in `parent` was already put in `destination` in `child`s dual list, they remain in the selected section. If I try to splice it from the selected, it just gets bounced back into `source`, even though when I look at the `source` contents, they don't contain the item... 

Any idea why this might be happening? 
If this is too complicated to understand without code example feel free to just close the issue.

**Comment by czeckd** [comment]
It's a fairly complicated scenario that I am not 100% certain I followed clearly.

From what you described, I understand this is you have this as your components:
![biskitcomp](https://user-images.githubusercontent.com/17859320/82980223-c6ab0500-9f9d-11ea-9383-fd80236aa29d.png)

Through some mechanism in the parent (outside of dual-list component) you are changing the allSourceOptions so that it does not have an option that ***is*** in the selectedSourceOptions and you do not get the results you want.

If my above understanding is correct, then I think I can explain why this happens. The dual-list does not move items from one array to the other array to select or deselect the items. So, if an item is not in the source, but is in the destination, them removal from the destination just removes it from the destination. It does not 'add' it to the source because it isn't there to begin with. For the removal to work and add it back to the source, for dual-list the source has to be a universal set of all items possible to add to the destination. 

From the [documentation](https://github.com/czeckd/angular-dual-listbox/blob/master/README.md): 

> * Technically, the dual-list component does not move items from one array to another. Rather it makes a copy from the source array of the item and adds it to the destination array, or removes it from the destination array. Thus, the source array is a master list of all available item and the destintion array is a list of items that have been selected from the master list. Therefore, in order for an item to be in the destination array it must also exist in the source array.

Another way to look at it, for the purposes of the dual list, the source is the master list.

`SourcePlanetsInSolarSystem: [ Mercury, Venus, Earth, Mars, Jupiter, Saturn, Uranus, Neptune ];`
`DestinationPlanetsVistedBySpaceProbe: [ Earth ];`

What is shown in the dual-list with these two arrays:

Source: Mercury, Venus, Mars, Jupiter, Saturn, Uranus, Neptune.
Destination: Earth

Note that `Earth` is in the `SourcePlanetsInSolarSystem` array. If the `Earth` was removed, it could still be in the `DestinationPlanetsVistedBySpaceProbe` array, however, if the `Earth` was removed from the dual-list destination array, then it would not appear in the `SourcePlanetsInSolarSystem` because its no longer there as a source.

Dual-list works differently than how most dual list controls function. It does not move/remove items from the source array. Rather it just hides/shows them as needed. If this isn't the problem, my apologies. I tried to understand your scenario.

TL/DR: All available options should be in the source for dual-list to function as designed.

**Comment by biskit1** [comment]
Thanks for taking the time to respond and explain. This is very helpful. Feel free to close issue :)
